node/src/main/scala/coop/rchain/node/instances/BlockProcessorInstance.scala

```diff
@@ -87,6 +89,13 @@ object BlockProcessorInstance {
                 _ <- logFinished
               } yield ()
             }
+            .evalTap { v =>
+              // we don't care about result of propose here, whether its started or not.
+              // so let deferred be not used
+              (Deferred[F, Option[Int]] >>= { d =>
+                proposerQueue.enqueue1((v._1, d))
+              }).whenA(autopropose)
```

Comment:
I'm not sure wdy mean, do you suggest to pass proposer instance inside block processor? I pass in only queue to minimize what I need to create BlockProcessorInstance, for better testability.

Reply:
This is all that is used here and on other places. 
```scala
def triggerPropose(casper: Casper[F]): F[Unit]
```

There is no need to push two arguments to a function and not use it. E.g why WebAPI needs to know about propose queue or auto propose flag if it's used there.

So what I mean is to create triggerPropose function in node runtime and supply it to the places where needed.

It would be even better that block API is instantiated with these data so that even less parameters are dragged through all these functions, but this is for future refactoring.

Reply:
I was thinking that the goal was to create streams/queues and connect different pieces of node together, so passing around queue for some worker entity (like proposer/block processor/etc..) as a generic principle is what we want.

Reply:
Streams should be used for lazy data processing and concurrent control flow. The purpose of Queues is to be a bridge between push and pull data/events e.g. transforming callbacks to Streams. And different pieces are connected with functions.

But I'm talking about more basic principle to not supply parameters to a function which are not used. Here is more simplified example. `a && b` represents duplicated code.
```scala
def doWork(start: Boolean) = ()

// Both modules use parameters exclusively together
def module1(a: Boolean, b: Boolean) = doWork(a && b)
def module2(a: Boolean, b: Boolean) = doWork(a && b)

// Common part can be extracted and supply to both modules
def module1a(c: Boolean) = doWork(c)
def module2a(c: Boolean) = doWork(c)

// Values
val A = true
val B = false

module1(A, B)
module2(A, B)

// Modules doesn't have to know about specific A and B if they are not used separately
val C = A & B
module1a(C)
module2a(C)
```

Reply:
I'm not sure I agree here - block processor should be informed, that under certain conditions it should send proposer a command to propose. For now it is called autopropose, but in future we can make it more granular, or maybe do tha after N block added or in some probabilistic manner. The thing is that `triggerPropose` should be aware where propose request come from, and then potentially have some logic inside. That imo better to be inside caller logic.

Reply:
Usually we want to make components not tightly coupled, which in this case I see Queue and creation of Deferred as internal state of proposer that is shared. I don't even see how other component can use Queue for other purposes then to enqueue elements, even worse if someone try to dequeue it, it will break proposer logic.
If there is additional logic in block processor it can skip calling triggerPropose and also triggerPropose can be extended with additional parameters if needs data from other components.

But I think you will agree that this code is duplicated on three places which can be extracted to a helper function.

https://github.com/rchain/rchain/blob/36cb8acb009ad4f3afd578a41ce1a5fe7ba8ce44/casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala#L70-L72

https://github.com/rchain/rchain/blob/36cb8acb009ad4f3afd578a41ce1a5fe7ba8ce44/node/src/main/scala/coop/rchain/node/instances/BlockProcessorInstance.scala#L95-L97

https://github.com/rchain/rchain/blob/36cb8acb009ad4f3afd578a41ce1a5fe7ba8ce44/node/src/main/scala/coop/rchain/node/instances/ProposerInstance.scala#L69-L71

Reply:
Just want to say I agree with @tgrospic on this.

Reply:
`even worse if someone try to dequeue it, it will break proposer logic` indeed, thanks, I will hide enqueuing behind some API